core/src/main/java/com/novoda/noplayer/exoplayer/ExoPlayerTwoImpl.java

```diff
@@ -0,0 +1,166 @@
+import com.novoda.noplayer.PlayerListenersHolder;
+import com.novoda.noplayer.PlayerView;
+import com.novoda.noplayer.Timeout;
+import com.novoda.noplayer.VideoContainer;
+import com.novoda.noplayer.VideoDuration;
+import com.novoda.noplayer.VideoPosition;
+import com.novoda.noplayer.player.PlayerInformation;
+
+import java.util.Collections;
+import java.util.List;
+
+public class ExoPlayerTwoImpl extends PlayerListenersHolder implements Player {
+
+    private final ExoPlayerTwoFacade facade;
+    private final Heart heart;
+    private final OnPrepareForwarder prepareForwarder;
+    private VideoContainer videoContainer = VideoContainer.empty();
+
+    public ExoPlayerTwoImpl(ExoPlayerTwoFacade facade) {
+        this.facade = facade;
+        Heart.Heartbeat<Player> onHeartbeat = new Heart.Heartbeat<>(getHeartbeatCallbacks(), this);
+        heart = Heart.newInstance(onHeartbeat);
+
+        prepareForwarder = new OnPrepareForwarder(getPreparedListeners(), this);
+        facade.addListener(prepareForwarder);
+        facade.addListener(new OnCompletionForwarder(getCompletionListeners()));
+        facade.addListener(new OnErrorForwarder(this, getErrorListeners()));
+        facade.addListener(new BufferStateForwarder(getBufferStateListeners()));
+        facade.addListener(new VideoSizeChangedForwarder(getVideoSizeChangedListeners()));
+
+        facade.setBitrateChangedListeners(getBitrateChangedListeners());
+        facade.setInfoListeners(getInfoListeners());
+    }
+
+    @Override
+    public boolean isPlaying() {
+        return facade.getPlayWhenReady();
+    }
+
+    @Override
+    public int getVideoWidth() {
+        return facade.getVideoWidth();
+    }
+
+    @Override
+    public int getVideoHeight() {
+        return facade.getVideoHeight();
+    }
+
+    @Override
+    public VideoPosition getPlayheadPosition() {
+        return VideoPosition.fromMillis(facade.getPlayheadPosition());
+    }
+
+    @Override
+    public VideoDuration getMediaDuration() {
+        return VideoDuration.fromMillis(facade.getMediaDuration());
+    }
+
+    @Override
+    public int getBufferPercentage() {
+        return facade.getBufferedPercentage();
+    }
+
+    @Override
+    public void play() {
+        showContainer();
+        heart.startBeatingHeart();
+        facade.setPlayWhenReady(true);
+        getStateChangedListeners().onVideoPlaying();
+    }
+
+    @Override
+    public void play(VideoPosition position) {
+        seekTo(position);
+        play();
+    }
+
+    @Override
+    public void pause() {
+        facade.setPlayWhenReady(false);
+        getStateChangedListeners().onVideoPaused();
+        if (heart.isBeating()) {
+            heart.stopBeatingHeart();
+            heart.forceBeat();
+        }
+    }
+
+    @Override
+    public void seekTo(VideoPosition position) {
+        facade.seekTo(position.inMillis());
+    }
+
+    @Override
+    public void reset() {
+        // TODO
```

Comment:
This TODO, unlike the others, was actually present in the old exo-player 😆  for the most part we ported over the TODOs from the exo-player one implementation.